### PR TITLE
Tutorial on multi-agent reinforcement learning with WarpDrive

### PR DESCRIPTION
**Pull request by sunil-s**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?

This PR adds a tutorial on multi-agent reinforcement learning with WarpDrive (https://github.com/salesforce/warp-drive) and Pytorch Lightning.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: lightning_examples/warp-drive/multi_agent_reinforcement_learning_with_warp_drive.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+num_gpus = 1
+assert num_gpus <= _NUM_AVAILABLE_GPUS, f"Only {_NUM_AVAILABLE_GPUS} GPU(s) are available!"
+num_epochs = (
+    run_config["trainer"]["num_episodes"]
+    * run_config["env"]["episode_length"]
+    / run_config["trainer"]["train_batch_size"]
+)
+
+trainer = Trainer(
+    gpus=num_gpus,
```
Comment (kaushikb11) [inline lightning_examples/warp-drive/multi_agent_reinforcement_learning_with_warp_drive.py:189]:
```suggestion
trainer = Trainer(
    accelerator="gpu",
    devices=num_gpus,
```

**Comment by Borda** [comment]
@krshrimali mind have a look ^^ :rabbit:

File: lightning_examples/warp-drive/multi_agent_reinforcement_learning_with_warp_drive.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+# %% [markdown]
+# # Learn More about WarpDrive and explore our tutorials!
+
+# %% [markdown]
+# 1. [WarpDrive basics](https://www.github.com/salesforce/warp-drive/blob/master/tutorials/tutorial-1-warp_drive_basics.ipynb)
+# 2. [WarpDrive sampler](https://www.github.com/salesforce/warp-drive/blob/master/tutorials/tutorial-2-warp_drive_sampler.ipynb)
+# 3. [WarpDrive reset and log](https://www.github.com/salesforce/warp-drive/blob/master/tutorials/tutorial-3-warp_drive_reset_and_log.ipynb)
+# 4. [Creating custom environments](https://www.github.com/salesforce/warp-drive/blob/master/tutorials/tutorial-4-create_custom_environments.md)
+# 5. [Training with WarpDrive](https://www.github.com/salesforce/warp-drive/blob/master/tutorials/tutorial-5-training_with_warp_drive.ipynb)
+# 6. [Scaling Up training with WarpDrive](https://www.github.com/salesforce/warp-drive/blob/master/tutorials/tutorial-6-scaling_up_training_with_warp_drive.md)
```
Comment (Borda) [inline lightning_examples/warp-drive/multi_agent_reinforcement_learning_with_warp_drive.py:380]:
```suggestion
```

File: .azure/ipynb-tests.yml [hunk 3]
```diff
@@ -17,8 +17,8 @@ jobs:
     # this need to have installed docker in the base image...
     container:
       # base ML image: mcr.microsoft.com/azureml/openmpi3.1.2-cuda10.2-cudnn8-ubuntu18.04
-      image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.8"
-      # image: "pytorch/pytorch:1.7.1-cuda11.0-cudnn8-runtime"
+      # image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.8"
+      image: "nvcr.io/nvidia/pytorch:22.01-py3"
```
Comment (Borda) [inline .azure/ipynb-tests.yml:21]:
```suggestion
      image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.8"
      # image: "pytorch/pytorch:1.7.1-cuda11.0-cudnn8-runtime"
```

File: .azure/ipynb-publish.yml [hunk 4]
```diff
@@ -21,8 +21,8 @@ jobs:
     # this need to have installed docker in the base image...
     container:
       # base ML image: mcr.microsoft.com/azureml/openmpi3.1.2-cuda10.2-cudnn8-ubuntu18.04
-      image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.8"
-      # image: "pytorch/pytorch:1.7.1-cuda11.0-cudnn8-runtime"
+      # image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.8"
+      image: "nvcr.io/nvidia/pytorch:22.01-py3"
```
Comment (Borda) [inline .azure/ipynb-publish.yml:25]:
```suggestion
      image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.8"
      # image: "pytorch/pytorch:1.7.1-cuda11.0-cudnn8-runtime"
```

**Comment by Emerald01** [comment]
@Borda @sunil-s 
Since warp_drive is a pure GPU lib, it is supposed to do all testing under GPU env so we have cuda compilers and other supporting stuffs, it will not pass a CPU test, as we discussed, maybe we need to address those things
- check why nbval on CPU is called while it shall be skipped
- find a way how to skip validate Docs  for this case
Thank you
Tian

**Comment by Borda** [comment]
@sunil-s is this some kind of artifact after package compilation?
```
LogicError: cuFuncSetBlockShape failed: invalid resource handle
```

**Comment by Emerald01** [comment]
> @sunil-s is this some kind of artifact after package compilation?
> 
> ```
> LogicError: cuFuncSetBlockShape failed: invalid resource handle
> ```

Look like a running error, this error comes from the failed CUDA kernel function, but usually it is caused by the error from some execution before it (since the error blocks the CUDA kernel from running).

**Comment by sunil-s** [comment]
> @sunil-s is this some kind of artifact after package compilation?
> 
> ```
> LogicError: cuFuncSetBlockShape failed: invalid resource handle
> ```

Hi @Borda , I just pulled the latest notebook and it runs fine on my CUDA machine: Python 3.7.2, PytorchLightning 1.6.1 and  Torch 1.9.0 Cuda 11.1
Perhaps, its still a version issue that's causing this?

**Comment by Borda** [comment]
> Python 3.7.2

is there a reason to limit it to this old python version?

**Comment by sunil-s** [comment]
> > Python 3.7.2
> 
> is there a reason to limit it to this old python version?

No reason, just that my machine has that version. If you would like me to test on a higher version, let me know.

File: lightning_examples/warp-drive/multi_agent_rl.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
+# %% [markdown]
+# # Fast Multi-agent Reinforcement Learning on a GPU using WarpDrive and Pytorch Lightning
+
```
Comment (Borda) [inline lightning_examples/warp-drive/multi_agent_rl.py:3]:
```suggestion
```
his comes from meta file

File: lightning_examples/warp-drive/multi_agent_rl.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+# We train the agents using $50$ environments or simulations running in parallel. With WarpDrive, each simulation runs on sepate GPU blocks.
+#
+# There are two separate policy networks used for the tagger and runner agents. Each network is a fully-connected model with two layers each of $256$ dimensions. We use the Advantage Actor Critic (A2C) algorithm for training. WarpDrive also currently provides the option to use the Proximal Policy Optimization (PPO) algorithm instead.
+
+# %%
+run_config = dict(
+    name="tag_continuous",
+    # Environment settings.
+    env=dict(
+        num_taggers=5,  # number of taggers in the environment
```
Comment (Borda) [inline lightning_examples/warp-drive/multi_agent_rl.py:61]:
```suggestion
        # number of taggers in the environment
        num_taggers=5,
```
I would do for all bellow so the notebook does ot have too long lines

File: lightning_examples/warp-drive/multi_agent_rl.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+    ax.set_xlim(0, 1)
+    ax.set_ylim(0, 1)
+    ax.set_zlim(-0.01, 0.01)
+
+    # Surface
+    corner_points = [
+        (0, 0),
+        (0, 1),
+        (1, 1),
+        (1, 0),
```
Comment (Borda) [inline lightning_examples/warp-drive/multi_agent_rl.py:176]:
```suggestion
        (1, 0)
```

File: lightning_examples/warp-drive/multi_agent_rl.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+        0,
+        0.02,
+        _get_label(0, init_num_runners, init_num_runners).lower(),
+    )
+
+    label.set_fontsize(14)
+    label.set_fontweight("normal")
+    label.set_color("#666666")
+
+    def animate(i):
```
Comment (Borda) [inline lightning_examples/warp-drive/multi_agent_rl.py:251]: consider move these function out of this parent func so it is not so heavy

File: lightning_examples/warp-drive/multi_agent_rl.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+# Define callbacks.
+cuda_callback = CudaCallback(module=wd_module)
+perf_stats_callback = PerfStatsCallback(
+    batch_size=wd_module.training_batch_size,
+    num_iters=wd_module.num_iters,
+    log_freq=log_freq,
+)
+
+# Instantiate the PytorchLightning trainer with the callbacks and the number of gpus.
+num_gpus = 1
```
Comment (Borda) [inline lightning_examples/warp-drive/multi_agent_rl.py:301]: duplicated to `_NUM_AVAILABLE_GPUS`
in fact I would remove the check `_NUM_AVAILABLE_GPUS` and set here fixed nb as 1

**Review by Borda — APPROVED** [review]

File: lightning_examples/warp-drive/multi_agent_rl.py (hunk 9 above)
Comment (sunil-s) [reply lightning_examples/warp-drive/multi_agent_rl.py:326]: Makes sense. Thanks. Resolved

File: lightning_examples/warp-drive/multi_agent_rl.py (hunk 6 above)
Comment (sunil-s) [reply lightning_examples/warp-drive/multi_agent_rl.py]: Done.

File: lightning_examples/warp-drive/multi_agent_rl.py (hunk 8 above)
Comment (sunil-s) [reply lightning_examples/warp-drive/multi_agent_rl.py:275]: Will leave it as is, and do no wish to expose this outside.

**Comment by Borda** [comment]
@kaushikb11 ^^

File: lightning_examples/warp-drive/multi_agent_rl.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+# %% [markdown]
+# ## Create the Lightning Trainer
+#
+# Next, we create the trainer for training the WarpDrive model. We add the `performance stats` callbacks to the trainer to view the throughput performance of WarpDrive.
+
+# %%
+log_freq = run_config["saving"]["metrics_log_freq"]
+
+# Define callbacks.
+cuda_callback = CudaCallback(module=wd_module)
```
Comment (kaushikb11) [inline lightning_examples/warp-drive/multi_agent_rl.py:317]:
```suggestion
cuda_callback = CUDACallback(module=wd_module)
```
Nit: Name it CUDACallback in main repo

File: lightning_examples/warp-drive/multi_agent_rl.py [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+import numpy as np
+import torch
+from example_envs.tag_continuous.tag_continuous import TagContinuous
+
+# from IPython.display import HTML
+from matplotlib import animation
+from matplotlib.patches import Polygon
+from pytorch_lightning import Trainer
+from warp_drive.env_wrapper import EnvWrapper
+from warp_drive.training.pytorch_lightning_trainer import CudaCallback, PerfStatsCallback, WarpDriveModule
```
Comment (kaushikb11) [inline lightning_examples/warp-drive/multi_agent_rl.py:32]:
```suggestion
from warp_drive.training.pytorch_lightning import CudaCallback, PerfStatsCallback, WarpDriveModule
```
Should probably change to this in the main repo

**Review by kaushikb11 — APPROVED** [review]

**Pull request merged**